tasm/tiles/src/LayoutDatabase.cc

```diff
@@ -0,0 +1,386 @@
+    std::string query = "INSERT INTO layouts (video, id, first_frame, last_frame, num_rows, num_cols)"\
+                    "VALUES (?, ?, ?, ?, ?, ?)";
+
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+    ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 3, firstFrame));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 4, lastFrame));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 5, numRows));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 6, numCols));
+
+    int result = sqlite3_step(select);
+    ASSERT_SQLITE_DONE(result);
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+}
+
+void LayoutDatabase::addTileLayoutWidths(const std::string &video, unsigned int id, const std::vector<unsigned int> &widths) {
+    std::string query = "INSERT INTO widths (video, id, width, position) "\
+                    "VALUES ( ?, ?, ?, ?)";
+
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+
+    int position = 0;
+    for (const auto& width : widths) {
+        ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 3, width));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 4, position));
+        ASSERT_SQLITE_DONE(sqlite3_step(select));
+        ASSERT_SQLITE_OK(sqlite3_reset(select));
+        ++position;
+    }
+
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+}
+
+void LayoutDatabase::addTileLayoutHeights(const std::string &video, unsigned int id, const std::vector<unsigned int> &heights) {
+    std::string query = "INSERT INTO heights (video, id, height, position)"\
+                    "VALUES (?, ?, ?, ?)";
+
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+
+    int position = 0;
+    for (const auto& height : heights) {
+        ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 3, height));
+        ASSERT_SQLITE_OK(sqlite3_bind_int(select, 4, position));
+        ASSERT_SQLITE_DONE(sqlite3_step(select));
+        ASSERT_SQLITE_OK(sqlite3_reset(select));
+        ++position;
+    }
+
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+}
+
+unsigned int LayoutDatabase::totalHeight(const std::string &video) const {
+    // select all ids for this video from layouts
+    std::string query = "SELECT id FROM layouts WHERE video = ?";
+    sqlite3_stmt *select;
```

Comment:
I think it's nice to check that the height is the same for all IDs, but I think it's something we can assume is true.  (Also, it will be super slow for long videos with lots of different layouts -- the complexity of this query will depend on the length of the video and number of layouts, rather than being constant). Then, a faster query would be "SELECT id FROM layouts WHERE video=? LIMIT 1".